### ISSUE TITLE
As a Client, Provide an Option to Use OAuth 2.0 Without OpenID

### ISSUE THREAD

**Issue by bschalme** [micronaut-projects/micronaut-security#507]
My experience with Micronaut 2.2.1 and Micronaut Security 2.2.0 is authenticating as a client using the OAUth 2.0 Authorization Code grant expects the authorization server to support OpenID. However, not all authorization servers support OpenID. For example, the [FreshBooks API](https://www.freshbooks.com/api/authentication) only deals with authorization codes, access tokens, refresh tokens, and (if you want) a state parameter. It does not support an OpenID issuer URL, nor does it return a JWT nor any other OpenID features.

It would be helpful if a client application using Micronaut Security could use OAuth 2.0 without OpenID, be it through a configuration setting in application.yml, or the Micronaut code gracefully accommodating the absence of OpenID settings in application.yml.

**Comment by sdelamo** [comment]
Our module already provides OAuth 2.0 support without OpenID Connect. 

See our [OAuth 2.0](https://micronaut-projects.github.io/micronaut-security/latest/guide/#oauth2-authorization-code) documentation. We have an example with Github which does not involve OpenID Connect. 

I am going to keep this issue open to investigate Freshbook API because if they don't use JWT you may need to provide a bean replacement.

**Comment by bschalme** [comment]
Ah hah! As I read section 15.3.1.1.2 (User Details Mapper) more carefully than I did previously, I see the first paragraph specifically calls out the requirement to define an implementation of OauthUserDetailsMapper. Once I did this, Micronaut was able to register the various OAuth 2.0 endpoints. There is no need for me to define an "opened" hierarchy in application.yml.

I am seeing Micronaut cobble together a JWT, rather than using one provided by FreshBooks. I simply defined a SecretSignatureConfiguration as described in the documentation.

Nevertheless, I am able to successfully get Micronaut Security to use the OAuth 2.0 Authorization Code grant to obtain an access token and a fresh token from FreshBooks. See [freshbooks-auth](https://github.com/bschalme/freshbooks-auth) for an example.

My apologies for the confusion. Thank you @sdelamo for your guidance and speedy response.

**Comment by sdelamo** [comment]
> I am seeing Micronaut cobble together a JWT, rather than using one provided by FreshBooks

You are probably using `cookie` authentication mode. To use the Freshbook access token directly you will need to do something like we are doing when we set `idtoken` authentication mode.

**Comment by bschalme** [comment]
Hmmm... FreshBooks' access token is just a randomly-generated 32-byte string. I'll see what I can do with `idtoken` authentication mode; I may need to JWT-ify their access token.

**Comment by sdelamo** [comment]
we should support flows which do not involve JWT tokens

**Comment by bschalme** [comment]
I got it to work!
I retained the `cookie` authentication mode, and ensured my `OauthUserDetailsMapper` populated the UserDetail's `attributes` map with the access and refresh tokens. Now a cookie named `JWT` is sent to the browser, and is used as the Bearer token in the Authorization header in subsequent requests. Recent updates to my repo [freshbooks-auth](https://github.com/bschalme/freshbooks-auth) shows this.

I would agree. FreshBooks is one, but I'm sure there are other providers who don't use JWTs and use the most basic form of OAuth 2.0.
